File: HoyDonde-frontend/services/securityAdminService.ts
import { apiClient } from './APIService';

/** Espejo de RolResponseDto (HoyDonde.API/DTOs). `codigo` es la identidad del documento: nunca editable. */
export interface RolResponse {
  codigo: string;
  nombre: string;
  descripcion: string;
  activo: boolean;
  acciones: string[];
}

/** Espejo de AccionResponseDto — `descripcion` es el nombre legible, `codigo` el técnico. */
export interface AccionResponse {
  codigo: string;
  descripcion: string;
  activo: boolean;
}

/** Espejo de UsuarioResumenResponseDto — nunca incluye ExternalSubjectId. */
export interface UsuarioResumenResponse {
  usuarioId: string;
  personaId: string;
  email: string;
  activo: boolean;
  rolesActivos: string[];
}

/** Espejo de PermisosEfectivosResponseDto — resuelto en vivo, nunca cacheado en el backend. */
export interface PermisosEfectivosResponse {
  usuarioId: string | null;
  personaId: string | null;
  usuarioActivo: boolean;
  roles: string[];
  acciones: string[];
}

export interface CreateRolPayload {
  codigo: string;
  nombre: string;
  descripcion: string;
}

export interface UpdateRolPayload {
  nombre: string;
  descripcion: string;
}

/**
 * Administración de roles/acciones/usuarios (`/api/security`, API_Documentation.md §10). El actor
 * siempre se resuelve en el backend desde el token — ningún método de acá manda un actor en el body.
 */
export const securityAdminService = {
  /** GET /api/security/roles — Policy ROL_EDITAR. */
  listRoles: async (): Promise<RolResponse[]> => {
    const response = await apiClient.get<RolResponse[]>('/security/roles');
    return response.data;
  },

  /** GET /api/security/roles/{codigo} — Policy ROL_EDITAR. Incluye las acciones asignadas. */
  getRol: async (codigo: string): Promise<RolResponse> => {
    const response = await apiClient.get<RolResponse>(`/security/roles/${codigo}`);
    return response.data;
  },

  /** POST /api/security/roles — Policy ROL_CREAR. */
  createRol: async (payload: CreateRolPayload): Promise<RolResponse> => {
    const response = await apiClient.post<RolResponse>('/security/roles', payload);
    return response.data;
  },

  /** PUT /api/security/roles/{codigo} — Policy ROL_EDITAR. Solo nombre/descripción: el código es inmutable. */
  updateRol: async (codigo: string, payload: UpdateRolPayload): Promise<RolResponse> => {
    const response = await apiClient.put<RolResponse>(`/security/roles/${codigo}`, payload);
    return response.data;
  },

  /** POST /api/security/roles/{codigo}/activar o /desactivar — Policy ROL_ACTIVAR. Idempotente. */
  setRolActivo: async (codigo: string, activo: boolean): Promise<void> => {
    await apiClient.post(`/security/roles/${codigo}/${activo ? 'activar' : 'desactivar'}`);
  },

  /** DELETE /api/security/roles/{codigo} — Policy ROL_ELIMINAR. Baja física: solo un rol personalizado, inactivo, sin usuarios asignados. */
  deleteRol: async (codigo: string): Promise<void> => {
    await apiClient.delete(`/security/roles/${codigo}`);
  },

  /** GET /api/security/acciones — Policy ROL_ASIGNAR_ACCION. Catálogo completo, siempre las mismas 20. */
  listAcciones: async (): Promise<AccionResponse[]> => {
    const response = await apiClient.get<AccionResponse[]>('/security/acciones');
    return response.data;
  },

  /** POST /api/security/roles/{rolCodigo}/acciones/{accionCodigo} — Policy ROL_ASIGNAR_ACCION. Idempotente. */
  asignarAccion: async (rolCodigo: string, accionCodigo: string): Promise<void> => {
    await apiClient.post(`/security/roles/${rolCodigo}/acciones/${accionCodigo}`);
  },

  /** DELETE /api/security/roles/{rolCodigo}/acciones/{accionCodigo} — Policy ROL_QUITAR_ACCION. Idempotente. */
  quitarAccion: async (rolCodigo: string, accionCodigo: string): Promise<void> => {
    await apiClient.delete(`/security/roles/${rolCodigo}/acciones/${accionCodigo}`);
  },

  /** GET /api/security/usuarios — Policy USUARIO_VER_PERMISOS_EFECTIVOS. */
  listUsuarios: async (): Promise<UsuarioResumenResponse[]> => {
    const response = await apiClient.get<UsuarioResumenResponse[]>('/security/usuarios');
    return response.data;
  },

  /** GET /api/security/usuarios/{usuarioId}/permisos-efectivos — Policy USUARIO_VER_PERMISOS_EFECTIVOS. */
  getPermisosEfectivos: async (usuarioId: string): Promise<PermisosEfectivosResponse> => {
    const response = await apiClient.get<PermisosEfectivosResponse>(`/security/usuarios/${usuarioId}/permisos-efectivos`);
    return response.data;
  },

  /** POST /api/security/usuarios/{usuarioId}/roles/{rolCodigo} — Policy USUARIO_ASIGNAR_ROL. Idempotente. */
  asignarRol: async (usuarioId: string, rolCodigo: string): Promise<void> => {
    await apiClient.post(`/security/usuarios/${usuarioId}/roles/${rolCodigo}`);
  },

  /** DELETE /api/security/usuarios/{usuarioId}/roles/{rolCodigo} — Policy USUARIO_QUITAR_ROL. Mismo guard del último Administrador. */
  quitarRol: async (usuarioId: string, rolCodigo: string): Promise<void> => {
    await apiClient.delete(`/security/usuarios/${usuarioId}/roles/${rolCodigo}`);
  },

  /** POST /api/security/usuarios/{usuarioId}/activar o /desactivar — Policy USUARIO_DESACTIVAR. Mismo guard del último Administrador. */
  setUsuarioActivo: async (usuarioId: string, activo: boolean): Promise<void> => {
    await apiClient.post(`/security/usuarios/${usuarioId}/${activo ? 'activar' : 'desactivar'}`);
  },
};

File: HoyDonde.API/Services/SecurityAdminService.cs
using HoyDonde.API.DTOs;
using HoyDonde.API.Exceptions;
using HoyDonde.API.Models;
using HoyDonde.API.Repositories;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace HoyDonde.API.Services
{
    // Administración configurable de roles/acciones asignadas/usuarios
    // (docs/security-refactor-plan.md §6, Etapa 5). Compone IRolRepository/IUsuarioRepository
    // (cada mutación transaccional con su SecurityAudit, ver esos repos) más la resolución de
    // permisos efectivos vía IPermissionService. El actor de cada operación se resuelve siempre
    // desde el externalSubjectId del token (nunca del body).
    public class SecurityAdminService : ISecurityAdminService
    {
        private static readonly Regex CodigoValido = new("^[A-Z][A-Z0-9_]*$", RegexOptions.Compiled);

        private readonly IRolRepository _rolRepository;
        private readonly IUsuarioRepository _usuarioRepository;
        private readonly IAccionRepository _accionRepository;
        private readonly IPermissionService _permissionService;
        private readonly IIdentityProvider _identityProvider;
        private readonly ISecurityAuditRepository _securityAuditRepository;

        public SecurityAdminService(
            IRolRepository rolRepository,
            IUsuarioRepository usuarioRepository,
            IAccionRepository accionRepository,
            IPermissionService permissionService,
            IIdentityProvider identityProvider,
            ISecurityAuditRepository securityAuditRepository)
        {
            _rolRepository = rolRepository;
            _usuarioRepository = usuarioRepository;
            _accionRepository = accionRepository;
            _permissionService = permissionService;
            _identityProvider = identityProvider;
            _securityAuditRepository = securityAuditRepository;
        }

        public async Task<RolResponseDto> CrearRolAsync(string actorExternalSubjectId, CreateRolRequestDto request)
        {
            ValidarCodigo(request.Codigo);
            if (string.IsNullOrWhiteSpace(request.Nombre)) throw new ArgumentException("El nombre del rol es obligatorio.");

            var actor = await ResolveActorAsync(actorExternalSubjectId);
            var rol = new Rol { Codigo = request.Codigo, Nombre = request.Nombre, Descripcion = request.Descripcion ?? string.Empty };

            var audit = NuevoAudit(actor, "ROL_CREAR", "Rol", request.Codigo,
                $"nombre={request.Nombre}");

            await _rolRepository.CrearAsync(rol, audit);
            return await ObtenerRolAsync(request.Codigo);
        }

        public async Task<RolResponseDto> EditarRolAsync(string actorExternalSubjectId, string codigo, UpdateRolRequestDto request)
        {
            if (string.IsNullOrWhiteSpace(request.Nombre)) throw new ArgumentException("El nombre del rol es obligatorio.");

            var actor = await ResolveActorAsync(actorExternalSubjectId);
            var audit = NuevoAudit(actor, "ROL_EDITAR", "Rol", codigo,
                $"nombre={request.Nombre}");

            await _rolRepository.EditarAsync(codigo, request.Nombre, request.Descripcion ?? string.Empty, audit);
            return await ObtenerRolAsync(codigo);
        }

        public async Task SetRolActivoAsync(string actorExternalSubjectId, string codigo, bool activo)
        {
            var actor = await ResolveActorAsync(actorExternalSubjectId);
            var audit = NuevoAudit(actor, activo ? "ROL_ACTIVAR" : "ROL_DESACTIVAR", "Rol", codigo, $"activo={activo}");

            await _rolRepository.SetActivoAsync(codigo, activo, audit);
        }

        public async Task EliminarRolAsync(string actorExternalSubjectId, string codigo)
        {
            var actor = await ResolveActorAsync(actorExternalSubjectId);
            var audit = NuevoAudit(actor, "ROL_ELIMINAR", "Rol", codigo, $"codigo={codigo}");

            await _rolRepository.EliminarAsync(codigo, audit);
        }

        public async Task<IReadOnlyList<RolResponseDto>> ListarRolesAsync()
        {
            var roles = await _rolRepository.GetAllAsync();
            var resultado = new List<RolResponseDto>();
            foreach (var rol in roles)
            {
                var acciones = await _rolRepository.GetAccionCodigosAsync(rol.Codigo);
                resultado.Add(MapRol(rol, acciones));
            }
            return resultado;
        }

        public async Task<RolResponseDto> ObtenerRolAsync(string codigo)
        {
            var rol = await _rolRepository.GetByCodigoAsync(codigo);
            if (rol == null) throw new RolNoEncontradoException(codigo);

            var acciones = await _rolRepository.GetAccionCodigosAsync(codigo);
            return MapRol(rol, acciones);
        }

        public async Task<IReadOnlyList<AccionResponseDto>> ListarAccionesAsync()
        {
            var acciones = await _accionRepository.GetAllAsync();
            return acciones.Select(a => new AccionResponseDto { Codigo = a.Codigo, Descripcion = a.Descripcion, Activo = a.Activo }).ToList();
        }

        public async Task<IReadOnlyList<UsuarioResumenResponseDto>> ListarUsuariosAsync()
        {
            var usuarios = await _usuarioRepository.GetAllAsync();
            return usuarios.Select(u => new UsuarioResumenResponseDto
            {
                UsuarioId = u.UsuarioId,
                PersonaId = u.PersonaId,
                Email = u.Email,
                Activo = u.Activo,
                RolesActivos = u.RolesActivos,
            }).ToList();
        }

        public async Task AsignarAccionARolAsync(string actorExternalSubjectId, string rolCodigo, string accionCodigo)
        {
            var actor = await ResolveActorAsync(actorExternalSubjectId);
            var audit = NuevoAudit(actor, "ROL_ASIGNAR_ACCION", "RolAccion", $"{rolCodigo}/{accionCodigo}",
                $"rol={rolCodigo};accion={accionCodigo}");

            await _rolRepository.AsignarAccionAsync(rolCodigo, accionCodigo, actor.UsuarioId, audit);
        }

        public async Task QuitarAccionDeRolAsync(string actorExternalSubjectId, string rolCodigo, string accionCodigo)
        {
            var actor = await ResolveActorAsync(actorExternalSubjectId);
            var audit = NuevoAudit(actor, "ROL_QUITAR_ACCION", "RolAccion", $"{rolCodigo}/{accionCodigo}",
                $"rol={rolCodigo};accion={accionCodigo}");

            await _rolRepository.QuitarAccionAsync(rolCodigo, accionCodigo, audit);
        }

        public async Task AsignarRolAUsuarioAsync(string actorExternalSubjectId, string usuarioId, string rolCodigo)
        {
            var rol = await _rolRepository.GetByCodigoAsync(rolCodigo);
            if (rol == null) throw new RolNoEncontradoException(rolCodigo);

            var actor = await ResolveActorAsync(actorExternalSubjectId);
            var audit = NuevoAudit(actor, "USUARIO_ASIGNAR_ROL", "UsuarioRol", $"{usuarioId}/{rolCodigo}",
                $"usuario={usuarioId};rol={rolCodigo}");

            await _usuarioRepository.AsignarRolAsync(usuarioId, rolCodigo, actor.UsuarioId, audit);
        }

        public async Task QuitarRolDeUsuarioAsync(string actorExternalSubjectId, string usuarioId, string rolCodigo)
        {
            var rol = await _rolRepository.GetByCodigoAsync(rolCodigo);
            if (rol == null) throw new RolNoEncontradoException(rolCodigo);

            var actor = await ResolveActorAsync(actorExternalSubjectId);
            var audit = NuevoAudit(actor, "USUARIO_QUITAR_ROL", "UsuarioRol", $"{usuarioId}/{rolCodigo}",
                $"usuario={usuarioId};rol={rolCodigo}");

            await _usuarioRepository.QuitarRolAsync(usuarioId, rolCodigo, audit);
        }

        public async Task<PermisosEfectivosResponseDto> ConsultarPermisosEfectivosAsync(string usuarioId)
        {
            var usuario = await _usuarioRepository.GetByIdAsync(usuarioId);
            if (usuario == null) throw new UsuarioNoEncontradoException(usuarioId);

            var permisos = await _permissionService.GetPermisosEfectivosPorUsuarioIdAsync(usuarioId);
            return new PermisosEfectivosResponseDto
            {
                UsuarioId = permisos.UsuarioId,
                PersonaId = permisos.PersonaId,
                UsuarioActivo = permisos.UsuarioActivo,
                Roles = permisos.Roles,
                Acciones = permisos.Acciones,
            };
        }

        public async Task SetUsuarioActivoAsync(string actorExternalSubjectId, string usuarioId, bool activo)
        {
            var actor = await ResolveActorAsync(actorExternalSubjectId);
            var audit = NuevoAudit(actor, activo ? "USUARIO_ACTIVAR" : "USUARIO_DESACTIVAR", "Usuario", usuarioId, $"activo={activo}");

            await _usuarioRepository.SetActivoAsync(usuarioId, activo, audit);
        }

        public async Task<PasswordResetLinkResponseDto> GenerarPasswordResetLinkAsync(string actorExternalSubjectId, string usuarioId)
        {
            var usuario = await _usuarioRepository.GetByIdAsync(usuarioId);
            if (usuario == null) throw new UsuarioNoEncontradoException(usuarioId);

            if (usuario.IdentityProvider != FirebaseIdentityProvider.ProviderName || string.IsNullOrEmpty(usuario.ExternalSubjectId))
            {
                throw new UsuarioSinIdentidadRecuperableException(usuarioId);
            }

            var actor = await ResolveActorAsync(actorExternalSubjectId);

            // Generar el enlace (Firebase, sistema externo) y auditar (Firestore) son dos escrituras
            // independientes -nunca una transacción distribuida-: si el proceso cae justo entre
            // ambas, el enlace ya fue emitido por Firebase pero la auditoría puede faltar. Nunca al
            // revés: si esta llamada falla, la ejecución nunca llega a auditar un éxito que no
            // ocurrió.
            var resetLink = await _identityProvider.GeneratePasswordResetLinkAsync(usuario.ExternalSubjectId);

            // Detalle deliberadamente vacío: nunca el email ni el enlace viajan a security_audits.
            var audit = NuevoAudit(actor, "USUARIO_GENERAR_RESET_PASSWORD", "Usuario", usuarioId, string.Empty);
            await _securityAuditRepository.RegistrarAsync(audit);

            return new PasswordResetLinkResponseDto { ResetLink = resetLink };
        }

        private static RolResponseDto MapRol(Rol rol, IReadOnlyList<string> acciones) => new()
        {
            Codigo = rol.Codigo,
            Nombre = rol.Nombre,
            Descripcion = rol.Descripcion,
            Activo = rol.Activo,
            Acciones = acciones.ToList(),
        };

        private static void ValidarCodigo(string? codigo)
        {
            if (string.IsNullOrWhiteSpace(codigo) || !CodigoValido.IsMatch(codigo))
            {
                throw new ArgumentException("El código del rol es inválido: debe ser no vacío, sin '/' ni caracteres de ruta, y usar solo mayúsculas, números y guion bajo (empezando con una letra).");
            }
        }

        private static SecurityAudit NuevoAudit((string UsuarioId, string PersonaId) actor, string operacion, string targetTipo, string targetId, string detalle) => new()
        {
            Id = Guid.NewGuid().ToString(),
            ActorUsuarioId = actor.UsuarioId,
            ActorPersonaId = actor.PersonaId,
            Operacion = operacion,
            TargetTipo = targetTipo,
            TargetId = targetId,
            Detalle = detalle,
            Timestamp = DateTime.UtcNow,
        };

        // El actor ya pasó por AccionAuthorizationHandler para llegar acá, así que su Usuario ya
        // está provisionado y activo; esta resolución es solo para completar los campos de
        // auditoría (ActorUsuarioId/ActorPersonaId), nunca para tomar una decisión de autorización.
        private async Task<(string UsuarioId, string PersonaId)> ResolveActorAsync(string externalSubjectId)
        {
            var usuarioId = await _usuarioRepository.GetUsuarioIdByExternalSubjectAsync(FirebaseIdentityProvider.ProviderName, externalSubjectId);
            if (usuarioId == null) throw new IdentityNotProvisionedException(externalSubjectId);

            var usuario = await _usuarioRepository.GetByIdAsync(usuarioId);
            if (usuario == null) throw new IdentityNotProvisionedException(externalSubjectId);

            return (usuario.Id, usuario.PersonaId);
        }
    }
}
